File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.cc
#include <bits/stdc++.h>
using namespace std;

string s;

bool check(string s) {
  stack<char> stk;
  for(char c : s) {
    if(c == '[' || c == '(') stk.push(c);
    else {
      if(c == ']') {
        if(stk.size() && stk.top() == '[') stk.pop();
        else return false;
      }
      if(c == ')') {
        if(stk.size() && stk.top() == '(') stk.pop();
        else return false;
      }
    }
  }
  return stk.empty();
}

int main() {
  while(true) {
    // 한 줄을 입력받아야 한다
    // cin >> s;
    getline(cin, s);
    if(s == ".") break;

    if(check(s)) cout << "yes\n";
    else cout << "no\n";
  }
  return 0;
}

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.js
const fs = require("fs");
// let input = fs.readFileSync("./input.txt");
let input = fs.readFileSync("/dev/stdin");

let sentences = input.toString().split("\n");

console.log(solution(sentences));
function solution(sentences) {
  const res = [];

  for (let i = 0; i < sentences.length; i++) {
    const sentence = sentences[i];
    if (sentence === ".") break;
    res.push(go(sentence));
    function go(sentence) {
      const stk = [];
      for (let j = 0; j < sentence.length; j++) {
        const char = sentence[j];
        if (char === "(" || char === "[") {
          stk.push(char);
        } else if (char === ")") {
          const top = stk[stk.length - 1];
          if (top === undefined) return "no";
          if (top !== "(") return "no";
          else stk.pop();
        } else if (char === "]") {
          const top = stk[stk.length - 1];
          if (top === undefined) return "no";
          if (top !== "[") return "no";
          else stk.pop();
        } else continue;
      }

      if (stk.length > 0) return "no";
      return "yes";
    }
  }

  return res.join("\n");
}
